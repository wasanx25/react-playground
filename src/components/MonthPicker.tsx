import React, { useEffect, useState } from 'react'

interface MonthPickerProps {
  handleChange: (yearAndMonth: string) => void
}

export const MonthPicker = (props: MonthPickerProps): JSX.Element => {
  const [isShow, setIsShow] = useState(false)
  const [year, setYear] = useState(2020)
  const [month, setMonth] = useState(12)

  useEffect(() => {
    props.handleChange(`${year}-${month}`)
  }, [year, month])

  function changeMonth(month: number) {
    setMonth(month)
    setIsShow(false)
  }

  return (
    <>
      <input type='text'
             placeholder='month picker'
             value={year + '-' + month}
             onBlur={_e => setIsShow(false)}
             onFocus={_e => setIsShow(true)}
             readOnly
      />

      <table className={isShow ? 'active' : 'inactive'}>
        <thead>
        <tr>
          <th onClick={_e => setYear(year - 1)}>{'<'}</th>
          <th>{year}</th>
          <th onClick={_e => setYear(year + 1)}>{'>'}</th>
        </tr>
        </thead>

        <tbody>
        <tr>
          {[1, 2, 3].map(value => {
            return <td key={value} onMouseDown={_e => changeMonth(value)}>{value}月</td>
          })}
        </tr>
        <tr>
          {[4, 5, 6].map(value => {
            return <td key={value} onMouseDown={_e => changeMonth(value)}>{value}月</td>
          })}
        </tr>
        <tr>
          {[7, 8, 9].map(value => {
            return <td key={value} onMouseDown={_e => changeMonth(value)}>{value}月</td>
          })}
        </tr>
        <tr>
          {[10, 11, 12].map(value => {
            return <td key={value} onMouseDown={_e => changeMonth(value)}>{value}月</td>
          })}
        </tr>
        </tbody>
      </table>

      <style jsx>{`
        .active {
          opacity: 1;
        }
        
        .inactive {
          opacity: 0;
        }

        table {
          position: absolute;
          z-index: 100;
          transition-property: opacity;
          transition-duration: .5s;
          background-color: #ffffff;
          margin: 1rem 0;
        }
        
        th:first-child:hover,
        th:last-child:hover,
        td:hover {
          background-color: #f5f3f3;
        }
        
        th, td {
          padding: .5rem 1rem;
          text-align: center;
        }
        
        th:first-child, th:last-child, td {
          cursor: pointer;
        }
      `}</style>
    </>
  )
}
