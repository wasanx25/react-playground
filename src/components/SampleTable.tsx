import React from "react";

export const SampleTable = (): JSX.Element => {
  return (
    <table>
      <caption>Sample Table</caption>
      <tbody>
        <tr>
          <th scope="rowgroup" rowSpan={3}>
            Group1
          </th>
          <th scope="row">RowHeader1</th>
          <td>Test1</td>
        </tr>
        <tr>
          <th scope="row">RowHeader2</th>
          <td>Test2</td>
        </tr>
        <tr>
          <th scope="row">RowHeader3</th>
          <td>Test3</td>
        </tr>
      </tbody>
    </table>
  );
};
