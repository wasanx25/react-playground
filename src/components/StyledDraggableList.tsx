import styled from '@emotion/styled'

export const StyledDraggableList = styled.ul(
  {
    '& li.dragover': {
      borderTop: 'solid 2px #bababa'
    }
  }
)

