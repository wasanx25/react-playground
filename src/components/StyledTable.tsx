import styled from "@emotion/styled";

export const Table = styled.table({
  position: "absolute",
  zIndex: 100,
  transition: "opacity 600ms, visibility 600ms",
  backgroundColor: "white",
  margin: "1rem 0",
  "&.active": {
    opacity: 1,
    visibility: "visible",
  },
  "&.inactive": {
    opacity: 0,
    visibility: "hidden",
  },
  "& .arrow:hover, & td:hover": {
    backgroundColor: "#f5f3f3",
  },
  "& th, & td": {
    textAlign: "center",
  },
  "& th": {
    padding: ".75rem 1rem",
  },
  "& td": {
    padding: ".5rem 1rem",
  },
  "& .arrow, & td": {
    cursor: "pointer",
  },
});
