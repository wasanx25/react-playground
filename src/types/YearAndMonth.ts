export class YearAndMonth {
  readonly year: string
  readonly month: string

  private constructor(year: string, month: string) {
    this.year = year
    this.month = month
  }

  static fromString(yearAndMonth: string): YearAndMonth {
    const [year, month] = yearAndMonth.split('-')
    return new YearAndMonth(year, month)
  }

  static now(): YearAndMonth {
    const date = new Date()
    const year = date.getFullYear()
    const month = date.getMonth() + 1

    return new YearAndMonth(year.toString(), month.toString())
  }

  format(): string {
    return `${this.year}-${this.month}`
  }
}
