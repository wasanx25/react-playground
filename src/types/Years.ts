import { eqDate } from "fp-ts/Date";
import { date } from "fp-ts";
import { pipe } from "fp-ts/function";
import { map, some } from "fp-ts/Option";
import { fold } from "fp-ts/boolean";

eqDate.equals(new Date(), new Date());
date.now();

pipe(
  some(true),
  map(
    fold(
      () => "false",
      () => "true"
    )
  )
);
