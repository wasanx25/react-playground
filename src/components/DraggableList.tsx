import React from 'react'
import { StyledDraggableList } from './StyledDraggableList'

const DRAGGABLE_LIST_ID = 'item-list'
const DATA_TRANSFER_ID = 'item-id'

export const DraggableList = (): JSX.Element => {
  function drag (event: React.DragEvent) {
    event.dataTransfer.setData(DATA_TRANSFER_ID, event.currentTarget.id)
  }

  function drop (event: React.DragEvent) {
    event.dataTransfer.dropEffect = 'move'

    const item = event.dataTransfer.getData(DATA_TRANSFER_ID)
    const element = document.getElementById(item)!
    const itemList = document.getElementById(DRAGGABLE_LIST_ID)!
    itemList.insertBefore(element, event.currentTarget)

    event.currentTarget.className = ''
  }

  function dragLeave (event: React.DragEvent) {
    event.currentTarget.className = ''
  }

  function dragOver (event: React.DragEvent) {
    event.preventDefault()
    event.currentTarget.className = 'dragover'
  }

  return (
    <>
      <StyledDraggableList id={DRAGGABLE_LIST_ID}>
        {[1, 2, 3, 4, 5].map(v => {
          return (
            <li key={v} draggable id={'drag-item-' + v} onDragStart={drag} onDrop={drop} onDragOver={dragOver}
                onDragLeave={dragLeave}>{v} Draggable Text
            </li>
          )
        })}
      </StyledDraggableList>
    </>
  )
}
