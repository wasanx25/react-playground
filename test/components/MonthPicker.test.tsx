import React from 'react'
import { render } from '@testing-library/react'
import { MonthPicker } from '../../src/components/MonthPicker'
import { YearAndMonth } from '../../src/types/YearAndMonth'
import '@testing-library/jest-dom/extend-expect'

describe('MonthPicker', () => {
  test.skip('sample', () => {
    const {getByRole} = render(<MonthPicker handleChange={jest.fn()} defaultValue={YearAndMonth.now()}/>)

    const expected = [
      '<thead>',
      '<tr>',
      '<th class="arrow">&lt;</th>',
      '<th>2020</th>',
      '<th class="arrow">&gt;</th>',
      '</tr>',
      '</thead>'
    ].join('')
    // expect(getByRole('month-picker-table', { hidden: true})).toMatchSnapshot()
    expect(getByRole('month-picker-table', {hidden: true})).toContainHTML(expected)
  })
})
