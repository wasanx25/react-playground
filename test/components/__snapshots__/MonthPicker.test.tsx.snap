// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthPicker sample 1`] = `
.c0 {
  position: absolute;
  z-index: 100;
  -webkit-transition: opacity 600ms,visibility 600ms;
  transition: opacity 600ms,visibility 600ms;
  background-color: #ffffff;
  margin: 1rem 0;
}

.c0.active {
  opacity: 1;
  visibility: visible;
}

.c0.inactive {
  opacity: 0;
  visibility: hidden;
}

.c0 .arrow:hover,
.c0 td:hover {
  background-color: #f5f3f3;
}

.c0 th,
.c0 td {
  text-align: center;
}

.c0 th {
  padding: .75rem 1rem;
}

.c0 td {
  padding: .5rem 1rem;
}

.c0 .arrow,
.c0 td {
  cursor: pointer;
}

<table
  class="c0 inactive"
  role="month-picker-table"
>
  <thead>
    <tr>
      <th
        class="arrow"
      >
        &lt;
      </th>
      <th>
        2020
      </th>
      <th
        class="arrow"
      >
        &gt;
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        1
        月
      </td>
      <td>
        2
        月
      </td>
      <td>
        3
        月
      </td>
    </tr>
    <tr>
      <td>
        4
        月
      </td>
      <td>
        5
        月
      </td>
      <td>
        6
        月
      </td>
    </tr>
    <tr>
      <td>
        7
        月
      </td>
      <td>
        8
        月
      </td>
      <td>
        9
        月
      </td>
    </tr>
    <tr>
      <td>
        10
        月
      </td>
      <td>
        11
        月
      </td>
      <td>
        12
        月
      </td>
    </tr>
  </tbody>
</table>
`;
